### PR TITLE
Make publish -r --since= great again

### LINKED ISSUE #1812: publish -r --since=EXACTCOMMIT goes into dirty submodules
So, even if specified a commit since which to "care", publish seems to go and "investigate" submodules which were not modified since last published point pointed by the commit in --since

```shell
$> git diff e1046e69133df9f6ab933a7759a64c2536b0ce0e.. --stat
 dbic | 2 +-
```
so, datalad should only consider publishing changes in `dbic` and this top level dataset, but log shows and it starts going through other submodules:

```shell
$> DATALAD_LOG_TIMESTAMP=1 datalad -l debug --dbg publish --to=datalad-public --missing=inherit -r --since=e1046e69133df9f6ab933a7759a64c2536b0ce0e
2017-09-08 13:27:45,314 [DEBUG  ] Determined class of decorated function: <class 'datalad.distribution.publish.Publish'>                           
2017-09-08 13:27:45,315 [DEBUG  ] Resolved dataset for publishing: <Dataset path=/mnt/btrfs/datasets/datalad/crawl> 
2017-09-08 13:27:45,334 [DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
2017-09-08 13:27:45,334 [DEBUG  ] Determined class of decorated function: <class 'datalad.interface.diff.Diff'> 
2017-09-08 13:27:45,334 [DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
2017-09-08 13:27:45,335 [DEBUG  ] Running: ['git', '--work-tree=.', 'diff', '--raw', '-z', '--ignore-submodules=none', '--abbrev=40', 'e1046e69133df9f6ab933a7759a64c2536b0ce0e', '--'] 
2017-09-08 13:27:49,163 [DEBUG  ] Running: ['git', '--work-tree=.', 'status', '--porcelain', '-z', '--ignore-submodules=all', '--untracked=all'] 
2017-09-08 13:27:49,180 [DEBUG  ] Determined class of decorated function: <class 'datalad.distribution.subdatasets.Subdatasets'> 
2017-09-08 13:27:49,180 [DEBUG  ] Resolved dataset for subdataset reporting/modification: <Dataset path=/mnt/btrfs/datasets/datalad/crawl> 
2017-09-08 13:27:49,181 [DEBUG  ] Running: ['git', '--work-tree=.', 'submodule', 'status'] 
2017-09-08 13:27:49,469 [DEBUG  ] Running: ['git', 'version'] 
2017-09-08 13:27:49,482 [DEBUG  ] Running: ['git', 'config', '-z', '-l', '--show-origin', '--file', '/mnt/btrfs/datasets/datalad/crawl/dbic/.datalad/config'] 
2017-09-08 13:27:49,495 [DEBUG  ] Running: ['git', 'config', '-z', '-l', '--show-origin'] 
2017-09-08 13:27:49,508 [DEBUG  ] Using existing Git repository at /mnt/btrfs/datasets/datalad/crawl/dbic 
2017-09-08 13:27:49,521 [DEBUG  ] Determined class of decorated function: <class 'datalad.distribution.subdatasets.Subdatasets'> 
2017-09-08 13:27:49,521 [DEBUG  ] Resolved dataset for subdataset reporting/modification: <Dataset path=/mnt/btrfs/datasets/datalad/crawl/dbic> 
2017-09-08 13:27:49,521 [DEBUG  ] Running: ['git', '--work-tree=.', 'submodule', 'status', '--recursive'] 
2017-09-08 13:27:49,578 [DEBUG  ] Determined class of decorated function: <class 'datalad.interface.diff.Diff'> 
2017-09-08 13:27:49,578 [DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
2017-09-08 13:27:49,579 [DEBUG  ] Running: ['git', '--work-tree=.', 'diff', '--raw', '-z', '--ignore-submodules=none', '--abbrev=40', 'e1046e69133df9f6ab933a7759a64c2536b0ce0e', '--'] 
2017-09-08 13:27:53,935 [DEBUG  ] Running: ['git', '--work-tree=.', 'status', '--porcelain', '-z', '--ignore-submodules=all', '--untracked=all'] 
2017-09-08 13:27:53,951 [DEBUG  ] Determined class of decorated function: <class 'datalad.interface.diff.Diff'> 
2017-09-08 13:27:53,952 [DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
2017-09-08 13:27:53,953 [DEBUG  ] Running: ['git', '--work-tree=.', 'diff', '--raw', '-z', '--ignore-submodules=none', '--abbrev=40', 'd89c835b0422f0df18454a8b538e41bf6bb75de7..3cceb1f9604e134628a56942210e43b74965bfe7', '--'] 
2017-09-08 13:27:54,040 [DEBUG  ] Using existing Git repository at /mnt/btrfs/datasets/datalad/crawl/dbic/QA 
2017-09-08 13:27:54,072 [DEBUG  ] Determined class of decorated function: <class 'datalad.interface.diff.Diff'> 
2017-09-08 13:27:54,072 [DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
2017-09-08 13:27:54,072 [DEBUG  ] Running: ['git', '--work-tree=.', 'diff', '--raw', '-z', '--ignore-submodules=none', '--abbrev=40', '9529efa47157e5e941381b22ae5bf8372befe74c..3006c6db5be73bedc4b67cf9649dfdc6761ee78c', '--'] 
2017-09-08 13:27:54,092 [DEBUG  ] Determined class of decorated function: <class 'datalad.distribution.subdatasets.Subdatasets'> 
2017-09-08 13:27:54,093 [DEBUG  ] Resolved dataset for subdataset reporting/modification: <Dataset path=/mnt/btrfs/datasets/datalad/crawl> 
2017-09-08 13:27:54,093 [DEBUG  ] Running: ['git', '--work-tree=.', 'submodule', 'status'] 
2017-09-08 13:27:54,359 [DEBUG  ] Running: ['git', 'version'] 
2017-09-08 13:27:54,370 [DEBUG  ] Running: ['git', 'config', '-z', '-l', '--show-origin', '--file', '/mnt/btrfs/datasets/datalad/crawl/kaggle/.datalad/config'] 
2017-09-08 13:27:54,381 [DEBUG  ] Running: ['git', 'config', '-z', '-l', '--show-origin'] 
2017-09-08 13:27:54,392 [DEBUG  ] Using existing Git repository at /mnt/btrfs/datasets/datalad/crawl/kaggle 
2017-09-08 13:27:54,408 [DEBUG  ] Determined class of decorated function: <class 'datalad.distribution.subdatasets.Subdatasets'> 
2017-09-08 13:27:54,408 [DEBUG  ] Resolved dataset for subdataset reporting/modification: <Dataset path=/mnt/btrfs/datasets/datalad/crawl/kaggle> 
2017-09-08 13:27:54,408 [DEBUG  ] Running: ['git', '--work-tree=.', 'submodule', 'status', '--recursive'] 
2017-09-08 13:27:54,464 [DEBUG  ] Determined class of decorated function: <class 'datalad.interface.diff.Diff'> 
2017-09-08 13:27:54,464 [DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
2017-09-08 13:27:54,465 [DEBUG  ] Running: ['git', '--work-tree=.', 'diff', '--raw', '-z', '--ignore-submodules=none', '--abbrev=40', 'e1046e69133df9f6ab933a7759a64c2536b0ce0e', '--'] 
2017-09-08 13:28:00,111 [DEBUG  ] Running: ['git', '--work-tree=.', 'status', '--porcelain', '-z', '--ignore-submodules=all', '--untracked=all'] 
2017-09-08 13:28:00,126 [DEBUG  ] Determined class of decorated function: <class 'datalad.interface.diff.Diff'> 
2017-09-08 13:28:00,127 [DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
2017-09-08 13:28:00,127 [DEBUG  ] Running: ['git', '--work-tree=.', 'diff', '--raw', '-z', '--ignore-submodules=none', '--abbrev=40', '82feee6097aea5c86f9919c32537212960809257', '--'] 
2017-09-08 13:28:00,155 [DEBUG  ] Running: ['git', '--work-tree=.', 'status', '--porcelain', '-z', '--ignore-submodules=all', '--untracked=all'] 
2017-09-08 13:28:00,169 [DEBUG  ] Determined class of decorated function: <class 'datalad.distribution.subdatasets.Subdatasets'> 
2017-09-08 13:28:00,169 [DEBUG  ] Resolved dataset for subdataset reporting/modification: <Dataset path=/mnt/btrfs/datasets/datalad/crawl> 
2017-09-08 13:28:00,169 [DEBUG  ] Running: ['git', '--work-tree=.', 'submodule', 'status'] 
2017-09-08 13:28:00,504 [DEBUG  ] Running: ['git', 'version'] 
2017-09-08 13:28:00,519 [DEBUG  ] Running: ['git', 'config', '-z', '-l', '--show-origin', '--file', '/mnt/btrfs/datasets/datalad/crawl/openfmri/.datalad/config'] 
2017-09-08 13:28:00,533 [DEBUG  ] Running: ['git', 'config', '-z', '-l', '--show-origin'] 
2017-09-08 13:28:00,553 [DEBUG  ] Using existing Git repository at /mnt/btrfs/datasets/datalad/crawl/openfmri 
2017-09-08 13:28:00,580 [DEBUG  ] Determined class of decorated function: <class 'datalad.distribution.subdatasets.Subdatasets'> 
2017-09-08 13:28:00,580 [DEBUG  ] Resolved dataset for subdataset reporting/modification: <Dataset path=/mnt/btrfs/datasets/datalad/crawl/openfmri> 
2017-09-08 13:28:00,581 [DEBUG  ] Running: ['git', '--work-tree=.', 'submodule', 'status', '--recursive'] 
2017-09-08 13:28:02,103 [DEBUG  ] Determined class of decorated function: <class 'datalad.interface.diff.Diff'> 
2017-09-08 13:28:02,103 [DEBUG  ] Determined class of decorated function: <class 'datalad.interface.annotate_paths.AnnotatePaths'> 
2017-09-08 13:28:02,104 [DEBUG  ] Running: ['git', '--work-tree=.', 'diff', '--raw', '-z', '--ignore-submodules=none', '--abbrev=40', 'e1046e69133df9f6ab933a7759a64c2536b0ce0e', '--'] 
^CERROR: 
Interrupted by user while doing magic:  [subprocess.py:_communicate_with_poll:1152]
2017-09-08 13:28:05,128 [DEBUG  ] Using existing Git repository at /mnt/btrfs/datasets/datalad/crawl 
DATALAD_LOG_TIMESTAMP=1 datalad -l debug --dbg publish --to=datalad-public  -  9.59s user 13.71s system 113% cpu 20.558 total


```

### PR DESCRIPTION
**Pull request by yarikoptic**
Yet another semi-blind fix on top of the previous tiny #1814 (since otherwise couldn't check) making it possible to just `publish -r --since=`.  The last commit is the trickiest since I think I violated some logic of tripple-annotating-and-diffing files whenever I do not care about files, but rather care about publishing changes to entire datasets.  I have changed to report that there is a diff if commits differ, which was the logic if no files were provided.  But the thing is that files were provided, since they were result of the first stage of annotating paths (why would we care about individual files when publishing full datasets?) and were passed below in the logic.

Another commit extended API of helpers and the interface to allow to not care about unstaged and not tracked files while annotating paths.  Made use of it in publish, since we can publish only what is committed (so why bother traversing into changes/untracked which aren't?).  Closes #1812  

Unfortunately only @mih could tell me where I violated anything, and now I am interested to see if tests pass.   I only know that `publish -r --since=` seems to work again.  I will give it a shot again after updating few datasets.

related closed whining: #1495

### CODE REVIEW

**Comment by bpoldrack** [comment]
FTR: test failure is the occasionally occuring one, which clearly is not related to this PR.

Looks reasonable to me. I guess the actual issue is to be sure about implications of #1814

File: datalad/distribution/publish.py [hunk 1]
```diff
@@ -98,12 +98,20 @@ def has_diff(ds, refspec, remote, paths):
...
-    if within_ds_paths:
+    commit_differ = current_commit != ds.repo.get_hexsha(remote_ref)
+    # yoh: not sure what "logic" was intended here for comparing only
+    # some files.  By now we get a list of files, if any were changed,
+    # from the commit on remote, and somehow diff says below that they didn't differ...
+    # but if commit is different -- there must be differences and we
+    # should publish. otherwise now skips publishing root dataset
+    # although its master is behind by 1 commit.  Moreover there could
+    # be an empty commit -- shouldn't we publish then???
+    if not commit_differ and within_ds_paths:
```
Comment (yarikoptic) [inline datalad/distribution/publish.py]: this is my main concern about my fix... not sure if this doesn't obliterate some magic, which should've worked while the fix had to be implemented elsewhere (e.g. in filtering the paths away, so here we would end up with no paths, thus pushing on irrespective according to the older logic)

Comment (yarikoptic) [reply datalad/distribution/publish.py]: So what do you think @mih?

**Comment by yarikoptic** [comment]
FWIW, just did `datalad publish -r --since= --missing=inherit --to=datalad-public` and it worked out splendidly publishing updates to openfmri and crcns

**Comment by yarikoptic** [comment]
@mih please give your blessing or input

**Comment by mih** [comment]
I found the word "splendid" in the discussion, so it must be good. I am lacking the mental capacity to fully grasp possible implications ATM. Let me just note that this adds a bunch of lines that are not executed in the tests. I fear that whatever works splendid now will not survive the next RF, and we will all see us on this island again ;-)

**Comment by yarikoptic** [comment]
pushed a slight extension to the test, hopefully captures a bit more.

**Comment by yarikoptic** [comment]
ok, seems to be a bit better now.  still few corner cases not even exercised though :-/ but since "splendid", will merge if no failures come up. Cheers
